### PR TITLE
#11859 Fix Read The Docs build process and update documentation deps.

### LINKED ISSUE #11859: Fix Read The Docs build
We a are buidling the docs with python 3.8 and some deps are no longer working  inside the RTD build environment.

### PR DESCRIPTION
**Pull request by adiroiban**
## Scope and purpose

Fixes #11859

We are building our docs with Python 3.8 and somehow the Read The Docs build environment no longer support this.

This PR tries to fix the RTD build.

It also tries to update the libraries that we use to build the documentation.

The goal is to build the docs using Sphinx 6 version.

The deps allow Sphinx v5 to support Python 3.7

This updated towncrier just to see that all is ok.
There is a separate ticket to use `towncrier` CLI for our CI, so that we don't have to reinvent the wheel https://github.com/twisted/twisted/issues/11583

This is also a separate ticket to cleanup pydoctor errors, so I think it's ok to just use latest pydocotor here  -  https://github.com/twisted/twisted/issues/11590

### CODE REVIEW

**Comment by adiroiban** [comment]
needs-review

**Comment by adiroiban** [comment]
The RTD looks ok

* https://twisted--11860.org.readthedocs.build/en/11860/
* API https://twisted--11860.org.readthedocs.build/en/11860/api/index.html

I see the checks are green, but there are failures

* sphinx doc build - https://github.com/twisted/twisted/actions/runs/4902054289/jobs/8753675388?pr=11860#step:5:165
* pydoctor apidic build - https://github.com/twisted/twisted/actions/runs/4902054289/jobs/8753674134?pr=11860#step:5:37


needs-changes

File: src/twisted/newsfragments/11859.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+The Read The Docs documentation build process was updated to latest Python and Sphinx version to fix a build error.
```
Comment (twm) [inline src/twisted/newsfragments/11859.bugfix]: I don't think that we really need this sort of build maintenance in the changelog.

**Review by twm — APPROVED** [review]
LGTM &mdash; I just made the same change in treq. Thank you @adiroiban!

**Comment by adiroiban** [comment]
Thanks, Tom for the review.

The Sphinx build now looks ok.

I have enabled auto-merge to get trunk green.

I hope there will be a new version of `cachecontrol` released soon ... but it might take some time for urllib3 support to be added. I see that `cachecontrol` is vendored in [pip](https://github.com/pypa/pip/tree/main/src/pip/_vendor) so I hope it will be fixed at some point soon

@glyph with this change, we pin urllib to an older version... but only for the `dev-release` extra.
Since this is not used for production deps, I hope this is ok

**Comment by glyph** [comment]
> @glyph with this change, we pin urllib to an older version... but only for the `dev-release` extra.
> Since this is not used for production deps, I hope this is ok

any pins to older versions are fine as long as we have a clear plan for getting out of them and documenting why they happen. even in the docs build, this will eventually break or have some version conflict with something else :).  Can you file a ticket on urllib, or cachecontrol, or sphinx, or wherever the real problem is, and link to it from an issue on Twisted so we will at least know when we can remove it?

File: pyproject.toml [hunk 2]
```diff
@@ -59,18 +59,22 @@ test = [
...
-    "readthedocs-sphinx-ext ~= 2.1",
-    "sphinx >= 4.1.2, <6",
+    "towncrier ~= 22.12",
+    "pydoctor ~= 23.4.0",
+    "sphinx-rtd-theme ~= 1.2",
+    "readthedocs-sphinx-ext ~= 2.2",
+    # We can move to v6 once we no longer need to support Python 3.7.
+    "sphinx >= 5, <7",
+    # This is due to cachecontrol not being compatible with urllib3>=2.0
+    # See https://github.com/ionrock/cachecontrol/issues/293
```
Comment (glyph) [inline pyproject.toml:69]: 👍 thanks, love to see it

**Pull request merged**